### PR TITLE
Use imported manifest for type inference (v2)

### LINKED ISSUE #69: React lib uses new cogs-client API

### PR DESCRIPTION
**Pull request by chetbox**
Resolves #69 

- use imported manifest for type inference
- Typescript 5.1 (fixes some type errors!)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
   "author": "Clockwork Dog <info@clockwork.dog>",
   "license": "MIT",
   "devDependencies": {
+    "@types/node": "18",
```
Comment (chetbox) [inline package.json:19]: Required because of `AbortSignal` type error with old `@types/node`

File: src/components/Hint.tsx [hunk 2]
```diff
@@ -1,9 +1,15 @@
 import { CogsConnection } from '@clockworkdog/cogs-client';
+import {} from '@clockworkdog/cogs-client/dist/types/ManifestTypes';
```
Comment (tomdixon) [inline src/components/Hint.tsx:2]: Left behind accidentally?

Comment (chetbox) [reply src/components/Hint.tsx]: Doh. Fixed.

File: package.json
Comment (tomdixon) [inline package.json:1]: Bump to version `2.0.0`?

File: src/hooks/useCogsEvent.ts [hunk 3]
```diff
@@ -1,22 +1,24 @@
-import { CogsConnection, EventKeyValue } from '@clockworkdog/cogs-client';
+import { CogsConnection, CogsIncomingEventTypes, ManifestTypes } from '@clockworkdog/cogs-client';
 import { useEffect } from 'react';
+import { ManifestFromConnection } from '../utils/types';
 
-type InputEventType<Connection> = Connection extends CogsConnection<infer Types> ? NonNullable<Types['inputEvents']> : never;
-
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export default function useCogsEvent<Connection extends CogsConnection<any>, EventKey extends keyof InputEventType<Connection>>(
+export default function useCogsEvent<
```
Comment (tomdixon) [inline src/hooks/useCogsEvent.ts:5]: In my testing of this, I'm trying to extract the `value` from the `event` with an Event type of `option` but the type always seems to be `undefined`

**Comment by tomdixon** [comment]
The standard pattern for a webpack setup so far has been to put the manifest file in the `public` folder so that it gets copied straight into the eventual `build` folder. This no longer works once you start importing the `.js` version as webpack doesn't like it being outside of the `src` folder.

I'm wondering what the neatest solution to this is?

**Comment by chetbox** [comment]
> The standard pattern for a webpack setup so far has been to put the manifest file in the `public` folder so that it gets copied straight into the eventual `build` folder. This no longer works once you start importing the `.js` version as webpack doesn't like it being outside of the `src` folder.
> 
> I'm wondering what the neatest solution to this is?

The neatest solution is to do the same thing and to import the manifest from your `public` folder:

```ts
import manifest from '../public/cogs-plugin-manfest.js`
```

I noticed Vite shows a warning in this case but it works just fine.

**Comment by tomdixon** [comment]
> > The standard pattern for a webpack setup so far has been to put the manifest file in the `public` folder so that it gets copied straight into the eventual `build` folder. This no longer works once you start importing the `.js` version as webpack doesn't like it being outside of the `src` folder.
> > I'm wondering what the neatest solution to this is?
> 
> The neatest solution is to do the same thing and to import the manifest from your `public` folder:
> 
> ```ts
> import manifest from '../public/cogs-plugin-manfest.js`
> ```
> 
> I noticed Vite shows a warning in this case but it works just fine.

From my tests I think with webpack this just doesn't compile?

**Comment by chetbox** [comment]
> > > The standard pattern for a webpack setup so far has been to put the manifest file in the `public` folder so that it gets copied straight into the eventual `build` folder. This no longer works once you start importing the `.js` version as webpack doesn't like it being outside of the `src` folder.
> > > I'm wondering what the neatest solution to this is?
> > 
> > 
> > The neatest solution is to do the same thing and to import the manifest from your `public` folder:
> > ```ts
> > import manifest from '../public/cogs-plugin-manfest.js`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I noticed Vite shows a warning in this case but it works just fine.
> 
> From my tests I think with webpack this just doesn't compile?

Did you set `allowJs` in your `tsconfig.json`? Maybe zip your project folder and post it here so I can take a look.

**Comment by tomdixon** [comment]
@chetbox This is essentially the cra-template project which I've been converting to use the new lib. If you `yarn build` you'll see the error. `allowJs` is `true`.

[react-proj.zip](https://github.com/clockwork-dog/cogs-client-react-lib/files/13165256/react-proj.zip)

**Comment by chetbox** [comment]
> @chetbox This is essentially the cra-template project which I've been converting to use the new lib. If you `yarn build` you'll see the error. `allowJs` is `true`.
> 
> [react-proj.zip](https://github.com/clockwork-dog/cogs-client-react-lib/files/13165256/react-proj.zip)

Ah this is a good find. Vite allows this and it works but the way Webpack is configured in create-react-app isn't happy.

I suggest making a symlink of `cogs-plugin-manifest.js` into the `public` folder or copying it to the `public` folder as part of the `yarn build` script. I can add this to the setup instructions. What do you think @tomdixon?

**Comment by tomdixon** [comment]
> > @chetbox This is essentially the cra-template project which I've been converting to use the new lib. If you `yarn build` you'll see the error. `allowJs` is `true`.
> > [react-proj.zip](https://github.com/clockwork-dog/cogs-client-react-lib/files/13165256/react-proj.zip)
> 
> Ah this is a good find. Vite allows this and it works but the way Webpack is configured in create-react-app isn't happy.
> 
> I suggest making a symlink of `cogs-plugin-manifest.js` into the `public` folder or copying it to the `public` folder as part of the `yarn build` script. I can add this to the setup instructions. What do you think @tomdixon?

Will that work on Windows?

**Comment by chetbox** [comment]
> > > @chetbox This is essentially the cra-template project which I've been converting to use the new lib. If you `yarn build` you'll see the error. `allowJs` is `true`.
> > > [react-proj.zip](https://github.com/clockwork-dog/cogs-client-react-lib/files/13165256/react-proj.zip)
> > 
> > 
> > Ah this is a good find. Vite allows this and it works but the way Webpack is configured in create-react-app isn't happy.
> > I suggest making a symlink of `cogs-plugin-manifest.js` into the `public` folder or copying it to the `public` folder as part of the `yarn build` script. I can add this to the setup instructions. What do you think @tomdixon?
> 
> Will that work on Windows?

Yes. Both Git and Windows support symlinks via [mklink](https://www.educative.io/answers/mklink-windows-10).

**Review by tomdixon — APPROVED** [review]

**Pull request merged**